lib/plugins/bed.js

```diff
@@ -24,17 +54,33 @@ function inject (bot) {
     } else if (bedBlock.type !== 26) {
       cb(new Error('wrong block : not a bed block'))
     } else {
+      const metadata = parseBedMetadata(bedBlock)
+      let headBlock = bedBlock
+
+      if (metadata.part === 0) {
+        headBlock = bot.blockAt(bedBlock.position.add(metadata.headOffset))
+        if (headBlock.type !== 26 || parseBedMetadata(headBlock).part === 0) {
+          return cb(new Error("there's only half bed"))
+        }
+      }
+
+      const topBlock = bot.blockAt(headBlock.position.offset(0, 1, 0))
+      if (topBlock.boundingBox === 'block') {
+        return cb(new Error("there's a block on top of the bed's head"))
+      }
+
       bot.activateBlock(bedBlock, () => {
+        function sleepCb () {
+          clearTimeout(sleepTimeout)
+          cb()
+        }
+
         const sleepTimeout = setTimeout(() => {
-          const sleepCb = bot.listeners('sleep').find(f => f.name === 'sleepCb')
           bot.removeListener('sleep', sleepCb)
           cb(new Error('there are monsters nearby'))
-        }, 500)
+        }, 1000)
```

Comment:
The problem will still exist, adding timeout won't help. On the other side 1s ping is kind of unplayable...

Reply:
@ImHarvol I opt for that solution you've proposed before. The thing is that with timeouts it's not that easy to do. Remember that you may lay in bed for entire night just because other players won't go to sleep. In that case any timeout fails.

But unfortunately some non-vanilla servers may disable this message or send a custom one